[addons][imagedecoder] rework about imagedecoder addon system

AlwinEsch opened the pull request:
## Description

In connection with my sandbox / API level project, some independent things that are necessary for this type of work are introduced here.

The support of image information is added as a new feature, with which the add-on can also give available information about Kodi.

It currently uses Kodi's unchanged CPictureInfoTag, but it would make sense to tackle and revise it soon.
Especially with CExifParse in it, this is very messy and flawed.
Think about changing it to [TinyEXIF](https://github.com/cdcseacave/TinyEXIF) which I now also use on imagedecoder.heif, this is much safer and more accurate than that in Kodi.

Also some tidying up to get it safer.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

![Bildschirmfoto vom 2021-11-12 17-00-55](https://user-images.githubusercontent.com/6879739/141497863-347004d3-0897-4ace-864c-a17f10d9d6c6.png)

![Bildschirmfoto vom 2021-11-06 15-54-13](https://user-images.githubusercontent.com/6879739/141497903-e7ee1a6f-9f55-4bc4-a6fe-615d57cb1477.png)
(ignore the Black/White in picture, was added in code after I created it)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][imagedecoder] pass "void*" value as addon target handle

The background is that in a sandbox system it is no longer possible to simply exchange a structure defined as a pointer between Kodi and addon, since Kodi never edits and uses this, it should always be passed as "void*" for something like that.

In addition, it becomes smaller because the target handle / class pointer add-on is then transferred directly and no longer the current API base structure.

AlwinEsch committed:
[addons][imagedecoder] clang cleanup about ImageDecoder.[cpp|h]

This to make it more clean on following changes.

AlwinEsch committed:
[addons][imagedecoder] remove "struct AddonProps_ImageDecoder" use

This structure was previously used to get used mime type. It comes into conflict with planned future ones, where it becomes more difficult to use it in all circumstances.
In addition, the creation of the new instance is a bit trickier because of the executed EXE type and where there would then be a problem with the structure so that the structure is to be transferred before the recipient is even there and ensures that it runs. Therefore, such types are only transferred there when the said instance is running.

With here a new callback function added to get used mime type from Kodi if needed.

Instead, requested values or the structure should be called up using direct functions.

AlwinEsch committed:
[addons][imagedecoder] rename ImageFormat to ADDON_IMG_FMT

This done to match his name as a correct "C" style enum. Also added ADDON_ to have more safe in case something else have same name.
Also does his enum name match then the begin of value names.

AlwinEsch committed:
[addons][imagedecoder] use typedef's for "C" API function pointers

The background to this is that this value is required in several places in the Sandbox API and is easier to manage using typedef.

It also makes the structure clearer to identify available functions.

I left the _V1 in there for the time being, because it would comb anyway and to leave out subsequent changes.
These commits and requests are mainly about having the finale cleaner and only recognizing the changes that are directly related to it.

AlwinEsch committed:
[addons][imagedecoder] add new system header (stdexcept) to API header

As far as possible, it must be possible to set up the header purely on "C" -based parent headers.

In addition, that it will be removed in the upcoming request to AddonBase.h (where currently provided), the background is because no throws are used there and can therefore be gotten cleaner.

AlwinEsch committed:
[addons][imagedecoder] change addon.xml extension/mimetype define way

This revises the way an image decoder communicates its supported formats.

Previously this was in a single line with "|" as a separator. If the list were to become very large, as is currently the case with the new support for text and image paths, it will no longer be transparent. Therefore it is now stored in its own XML group.

Also includes this the SupportsFile function with where addon can ask about support of wanted file.

AlwinEsch committed:
[addons][imagedecoder] change used namespace from ADDON to KODI::ADDONS

This cleans up the used namespace of the two files using the new style which should now be used.

In future, all of them in the "xbmc/addons" folder should be converted to this.

AlwinEsch committed:
[addons][imagedecoder] make KodiToAddonFormat in pseudo namespace and use as constexpr

This corrects the use of KodiToAddonFormat in ImageDecoder.cpp. Before it was not in a namespace, which could be problematic.
In addition, what it is set as std::map, here it is changed to std::array and with constexpr, which requires less computing work and memory.

AlwinEsch committed:
[addons][imagedecoder] Read color size from KodiToAddonFormat

Previously this was done using switch / case, but since there is already a list, it is now moved to there and done together with the conversion of the Kodi format to the add-on format.

In addition, the values have been changed to "const" because there are no more changes to them anyway.

AlwinEsch committed:
[addons][imagedecoder] rename image_decoder.h to imagedecoder.h

With the newly planned sandbox / API level system it becomes necessary to have the "C" files matching its namespace.

As an example, a "C" function in imagedecoder must use "kodi_addons_imagedecoder_some_function ()" as the name, with which the "_" acts as a namespace divider to begin with, if now, as before, it would see "image" and "decoder" as separate areas.

Advantageous for the automatic creation script, otherwise problems can arise for him.

AlwinEsch committed:
[addons][imagedecoder] Only pass mimetype by LoadImageFromMemory

Previously it was used as a callback (or as a property value). However, since this value is only available when processing the decoding, it makes no sense to have a value that is always available via the callback.

This removes it and is only available with the LoadImageFromMemory to the addon.

AlwinEsch committed:
[addons][imagedecoder] Add Picture InfoTag support

This changes the image encoder addon interface with the addition of a new function to retrieve image information.

The cool effect is that the performance is significantly improved, since Kodi now creates the cached and shortcut icons of the images when loading "tag". As a result, the open window is much faster than before. In addition, there is a GUI process in the loading time of infos, which makes it better for the user.

The info day in Kodi is not yet fully used by add-ons and is inserted as a comment in "CImageDecoder", but first has to look through where, how and when it is used. With recent changes, a large part is created in connection with imagedecoder.raw addon.
This may be expanded in the future.

AlwinEsch committed:
[addons][imagedecoder] make use more safe in case of error

This removes the `Create()` function in CImageDecoder and executes it in constructor. To check for a correct start, `IsCreated()` has been added.

The `m_created` is also checked for all calls in order to always avoid being used after a failed initialization.

AlwinEsch committed:
[addons][imagedecoder] make static API function names on C++ header equal to C

This change has to keep the function names of the "C" part the same as the static am on the C++, only "ADDON_" remains as the beginning, because I suspected that with e.g. "init" there could be problems in other places.

Another reason is the planned future automation of the interface between Kodi and addon, where it will be automatically added to the C ++ header. Hence, names must be kept the same.

AlwinEsch committed:
[addons][imagedecoder] increase API version to 3.0.0

In connection with previous commits.

garbear commented on xbmc/addons/kodi-dev-kit/include/kodi/c-api/addon-instance/imagedecoder.h:
```diff
@@ -0,0 +1,451 @@
+  /// Kodi.
+  ///
+  ///@{
+  typedef enum ADDON_IMG_FMT
```
Are there other image formats possible? We define pixel layouts quite often, is it possible to use or expand an existing pixel layout enum?

garbear replied:
Thinking about it, it's probably better to separate image pixel formats from other pixel formats, as images have different requirements than video of games for example. So disregard.

AlwinEsch replied:
> Are there other image formats possible? We define pixel layouts quite often, is it possible to use or expand an existing pixel layout enum?

Not really the enum is only in relationship to Kodi's one here: https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/TextureFormats.h#L18_L21, any others are currently not there (them below there a fixed to Windows and unusable by addon).

Them on game by `GAME_PIXEL_FORMAT` relates to VideoPlayers `AVPixelFormat`.

> Thinking about it, it's probably better to separate image pixel formats from other pixel formats, as images have different requirements than video of games for example. So disregard.

:+1:

garbear commented:
I'll get grammar improvements for the string descriptions in a bit. The comments are mostly fine, do you still want recommendations there?

garbear commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -13861,7 +13861,19 @@ msgctxt "#21345"
+#. Used to give heading list entry for an mimetype description, e.g. to show as 'Extension: "audio/mpeg3"'
+#: xbmc/addons/GUIDialogAddonInfo.cpp
+msgctxt "#21347"
+msgid "Mimetype: \"{0:s}\""
```
Correct capitalization when used in a sentence is "MIME type"

garbear commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -13861,7 +13861,19 @@ msgctxt "#21345"
 msgstr ""
 
-#empty strings from id 21346 to 21357
+#. Used to give heading list entry for an file extension description, e.g. to show as 'Extension: ".mp3"'
```
I'll get grammar improvements for the string descriptions in a bit. The comments are mostly fine, do you still want recommendations there?

garbear commented:
I read from top to bottom and didn't see any other issues. Next pass I'll go through commit by commit.

AlwinEsch committed:
[addons][imagedecoder] minor addon.xml doc change

This corrects the documentation about addon.xml for imagedecoder addons and to add new used ways to declare supported mimetype and extension.

AlwinEsch committed:
[addons][imagedecoder] use size_t for reading data on array.

This is a mandatory guideline on the revised API!

It must never be used as such `uint8_t* buffer, int size` or like here as `unsigned char*` alone, the autogeneration can then not identify it as array and thing only about single value where can be write. it must always be set with a size_t after its value.

enen92 commented on xbmc/addons/ExtsMimeSupportList.cpp:
```diff
@@ -28,6 +28,14 @@ CExtsMimeSupportList::CExtsMimeSupportList(CAddonMgr& addonMgr) : m_addonMgr(add
+
+  addonInfos.clear();
+
+  if (m_addonMgr.GetAddonInfos(addonInfos, true, ADDON_IMAGEDECODER))
```
Maybe it makes sense to overload GetAddonInfos to accept a list of types to return (or a variable number of arguments)? Seems unnecessary to request the list and iterate it twice for each type

AlwinEsch replied:
Would be the best in itself, just not sure whether the overkill of work would be worth it.

A new function could be introduced to support this with vector. Expanding the current one https://github.com/xbmc/xbmc/blob/master/xbmc/addons/AddonManager.cpp#L1160-L1175 would be bad for me, as 99% only need one type and then there is a lot of unnecessary work.

Another way would be to use ADDON_UNKNOWN, but this would transfer all available addons, which would also be more data.

AlwinEsch replied:
Will see to use it as an independent function (without using or changing an existing one)

AlwinEsch replied:
@enen92 have added a new commit where adds a new function on addon manager to get infos by type selection list.

AlwinEsch committed:
[addons] allow get of CAddonInfo's by selected types list

This introduces a new function to call up several types of add-ons in a list.

Also reduced the code on "CExtsMimeSupportList::CExtsMimeSupportList".

AlwinEsch commented:
If nothing comes I bring in in one hour and to update tommorow the binary addons (after Ubuntu PPA updated).

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
As merging is blocked now, can someone please review and approve or decline the request.
